### DISCUSSION TITLE
Gleam Style Guide

### DISCUSSION THREAD

**Discussion by NNBnh** [gleam-lang/gleam#3785] [Ideas & suggestions]
Hello, Gleam is my most favorite programing language! But while I learning it, I just fell kind of unsure if the way I write my code, the name I named values and functions are truly correct or optimal. I want to know the DO and DON'T of Gleam, it's would be nice to have a Gleam Style Guide like [Ruby Style Guide](https://rubystyle.guide), [Godot Style Guide](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html) or [Rust Style Guide](https://doc.rust-lang.org/nightly/style-guide) that help newcomer to improve their skill with Gleam and also set a standard so that Gleam code across projects and teams can be more consistent.

E.g: How do you name the function that do internal recursion stuffs for a interface function that other actually call? I saw it got postfix with `_loop` in one of [Gleam tour example](https://tour.gleam.run/flow-control/tail-calls), but is this the standard?

```gleam
import gleam/io

pub fn main() {
  io.debug(factorial(5))
  io.debug(factorial(7))
}

pub fn factorial(x: Int) -> Int {
  // The public function calls the private tail recursive function
  factorial_loop(x, 1)
}

fn factorial_loop(x: Int, accumulator: Int) -> Int {
  case x {
    0 -> accumulator
    1 -> accumulator

    // The last thing this function does is call itself
    // In the previous lesson the last thing it did was multiply two ints
    _ -> factorial_loop(x - 1, accumulator * x)
  }
}
```

**Comment by GearsDatapacks** [comment]
Often `do_factorial` would be used here

**Reply by lpil** [reply]
While that's a convention in the stdlib I wouldn't say it's something that's encouraged or not in the wider ecosystem

**Comment by lpil** [comment]
From my point of view between the highly constrained nature of Gleam and the inclusion of a formatter there's not need for a style guide like this. The idea was that rather than having a large set of rules to nit-pick the environment should result in your writing code that would already conform to the desired style. This should overall provide a much more comfortable programming experience.

That said, it's not a perfect system, so there's still things to advise, likely at a higher level than those guides. I intend to write more about how to write Gleam code next year.

**Comment by inoas** [comment]
Semi-OT: The _loop suffix seems great for people coming to FP and wondering how to loop not knowing much about vanilla recursion (or tail recursion)

**Reply by lpil** [reply]
Yes I think I agree. Better than do_
